raiden_contracts/contracts/TokenNetwork.sol

```diff
@@ -275,296 +278,299 @@ contract TokenNetwork is Utils {
-
-        // The closer is not allowed to call updateTransfer
-        require(closing_requests[channel_identifier].closing_participant != msg.sender);
-
-        // Call the private function for the actual logic and constraints of updateTransfer
-        updateTransferPrivate(
-            channel_identifier,
-            nonce,
-            transferred_amount,
-            locksroot,
-            additional_hash,
-            closing_signature
-        );
-    }
-
     /// @notice Called on a closed channel, the function allows the non-closing participant to
     // provide the last balance proof, which modifies the closing participant's state. Can be
-    // called multiple times by anyone, as long as they provide signatures from both participants.
+    // called multiple times by anyone.
+    /// @param balance_hash Hash of (transferred_amount, locked_amount, locksroot).
+    /// @param additional_hash Computed from the message. Used for message authentication.
     /// @param channel_identifier The channel identifier - mapping key used for `channels`.
     /// @param nonce Strictly monotonic value used to order transfers.
-    /// @param transferred_amount Total amount of tokens transferred by the channel partner
-    /// to the channel participant who calls the function.
-    /// @param locksroot Root of the partner's merkle tree of all pending lock lockhashes.
-    /// @param additional_hash Computed from the message. Used for message authentication.
     /// @param closing_signature Closing participant's signature of the balance proof data.
     /// @param non_closing_signature Non-closing participant signature of the balance proof data.
-    function updateTransferDelegate(
+    function updateNonClosingBalanceProof(
         uint256 channel_identifier,
-        uint64 nonce,
-        uint256 transferred_amount,
-        bytes32 locksroot,
+        bytes32 balance_hash,
+        uint256 nonce,
         bytes32 additional_hash,
         bytes closing_signature,
         bytes non_closing_signature)
         external
     {
-        // We also need the signature from the non-closing participant on behalf of which the 3rd
-        // party makes the transaction. This signature will be provided to the 3rd party
-        //  together with the balance proof.
+        Channel storage channel = channels[channel_identifier];
+
+        // Channel must be closed
+        require(channel.state == 2);
+
+        // Channel must be in the settlement window
+        require(channel.settle_block_number >= block.number);
+
+        // We need the signature from the non-closing participant to allow anyone
+        // to make this transaction. E.g. a monitoring service.
         address non_closing_participant = recoverAddressFromBalanceProof(
             channel_identifier,
+            balance_hash,
             nonce,
-            transferred_amount,
-            locksroot,
             additional_hash,
```

Comment:
I did not see the point in including the `closing_signature` here. 
Not including it keeps the code more dry (1 function for all balance proof address recoveries)

Reply:
Both signatures are **required**. The first signature proves the balance proof was sent by the channel participant and it's valid, the second proves the balance proof was validated and allows delegate calls. If the closing signature is not included this attack is possible:

- The attacker makes sure to **never** send a balance proof with a nonce higher than the **counter party's** *latest* balance proof, **and** to exhaust *its* capacity from the channel (this is really easy to do, just wait for the counter party to use the channel twice and send a single payment that uses all the available balance).
- Then the attacker closes the channel. It will provide the **latest valid** balance proof from the counter party to properly close it.
- Then the attacker gets the counter party balance proof with a nonce *equal* to **its** latest balance proof nonce + 1, signs it, and calls `updateNonClosingBalanceProof` like it was a monitoring service. The balance proof and signature are **valid**, but for the **wrong direction**, and because the attacker took care to never produce a balance proof with a higher nonce the other node cant overwrite it.

The **wrapping** of the signatures is what proves the direction of the balance proof.

Reply:
Ok, yes, agreed. Will be changed.

Reply:
Fixed here https://github.com/raiden-network/raiden-contracts/pull/39/commits/40388b30e6daeb1efc55304f08f4e60d90ef342a